fvm/state/state.go

```diff
@@ -79,132 +84,210 @@ func WithMaxInteractionSizeAllowed(limit uint64) func(st *State) *State {
 	}
 
-	s.draft[fullKey(owner, controller, key)] = payload{owner, controller, key, value}
+	pKey := PayloadKey{owner, controller, key}
+	p := Payload{pKey, value}
+	s.LogTouch(&p)
+
+	s.updateDelta(&p)
+
 	address, isAddress := addressFromOwner(owner)
 	if isAddress {
 		s.updatedAddresses[address] = struct{}{}
 	}
 	return nil
 }
 
-func (s *State) Touch(owner, controller, key string) error {
-	_, err := s.Get(owner, controller, key)
-	return err
-}
-
 func (s *State) Delete(owner, controller, key string) error {
 	err := s.Set(owner, controller, key, nil)
 	return err
 }
 
-func (s *State) Commit() error {
-	s.lock.Lock()
-	defer s.lock.Unlock()
+// We don't need this later, it should be invisible to the cadence
+func (s *State) Touch(owner, controller, key string) error {
+	s.LogTouch(&Payload{PayloadKey{owner, controller, key}, nil})
+	return nil
+}
+
+// NewChild generates a new child state
+func (s *State) NewChild() *State {
+	return NewState(s.ledger,
+		WithParent(s),
+		WithMaxKeySizeAllowed(s.maxKeySizeAllowed),
+		WithMaxValueSizeAllowed(s.maxValueSizeAllowed),
+		WithMaxInteractionSizeAllowed(s.maxInteractionAllowed),
+	)
+}
+
+func (s *State) MergeTouchLogs(child *State) error {
+	// append touches
+	s.LogTouches(child.touchLog)
+	// TODO maybe merge read cache for performance on failed cases
+	return nil
+}
 
-	// TODO right now we sort keys to minimize
-	// the impact on the spock, but later we might
-	// expose interactionFingerprint as a separate
-	// parameters, this preserve number of reads and
-	// order of reads as extra layer of entropy
+// MergeAnyState applies the changes from any given state
+func (s *State) MergeAnyState(other *State) error {
+	// append touches
+	s.LogTouches(other.touchLog)
 
-	keys := make([]string, 0)
-	for k := range s.draft {
-		keys = append(keys, k)
+	// apply delta
+	for _, v := range other.delta {
+		s.updateDelta(&v)
 	}
 
-	sort.Strings(keys)
+	// apply address updates
+	for k, v := range other.updatedAddresses {
+		s.updatedAddresses[k] = v
+	}
 
-	for _, k := range keys {
-		p := s.draft[k]
-		oldValue, err := s.ledger.Get(p.owner, p.controller, p.key)
-		if err != nil {
-			return err
-		}
-		// update interaction
-		err = s.updateInteraction(p.owner, p.controller, p.key, oldValue, p.value)
+	// update ledger interactions
+	s.ReadCounter += other.ReadCounter
+	s.WriteCounter += other.WriteCounter
+	s.TotalBytesRead += other.TotalBytesRead
+	s.TotalBytesWritten += other.TotalBytesWritten
+
+	// check max interaction as last step
+	return s.checkMaxInteraction()
+}
+
+// MergeState applies the changes from a the given view to this view.
+func (s *State) MergeState(child *State) error {
+	// merge read cache
+	for k, v := range child.readCache {
+		s.readCache[k] = v
+	}
+
+	return s.MergeAnyState(child)
+}
+
+// ApplyDeltaToLedger should only be used for applying changes to ledger at the end of tx
+// if successful
+func (s *State) ApplyDeltaToLedger() error {
+	for _, v := range s.delta {
```

Comment:
maybe we could check that s.parent is nil here. This should only be done on the root right?

Reply:
I removed this logic completely so we don't need this function